### PR TITLE
WIP: Fix/improve check on ce

### PR DESCRIPTION
**Pull request by thepassle**
Doesnt quite work yet. 

from the extension `el.constructor !== HTMLElement === false`
Screenshot 2020-09-04 at 16 56 52

But in the browser its `true`:
Screenshot 2020-09-04 at 16 57 17

### CODE REVIEW

**Comment by jlengrand** [comment]
Just wondering, is it the right moment to start adding a test or two for the checks, or is it too early? I see boolean returns based on three conditions and my spider sense starts tickling

File: src/content_script.js [hunk 1]
```diff
@@ -7,7 +7,7 @@ let index = 0;
 
 function isCustomElement(el) {
   const isAttr = el.getAttribute('is');
-  return el.localName.includes('-') || isAttr && isAttr.includes('-');
+  return el.constructor !== HTMLUnknownElement && el.constructor !== HTMLElement && el.localName.includes('-') || isAttr && isAttr.includes('-');
```
Comment (jorgecasar) [inline src/content_script.js]: It's not easer something like this:

```
function isCustomElement(el) {
   return window.customElements.get(el.nodeName.toLowerCase());
}
```

File: package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "custom-elements-locator",
-  "version": "1.3.0",
+  "version": "1.4.0",
```
Comment (jorgecasar) [inline package.json]: Remember to generate package-lock with the same version

File: src/content_script.js (hunk 1 above)
Comment (thepassle) [reply src/content_script.js]: That would be easier, but the content script is sandboxed, window.customElements.get will fail 🤷

**Comment by thepassle** [comment]
Looks like `el.constructor !== HTMLElement` is always `false` because in a contentscript you dont actually have access to the pages JS/variables 🤦  you only have trivial access to the dom, and since you dont have access to the pages JS, it'll just say its HTMLElement 🤷 

for the same reason you cant do `customElements.get()` in a contentscript

Need to think a little more on this, would be nice to find out if theres another way to find out if a dom node is a real custom element

**Comment by jorgecasar** [comment]
I guess I got the solution. Try with `el.constructor.__proto__` instead of `el.constructor`

**Comment by thepassle** [comment]
> I guess I got the solution. Try with `el.constructor.__proto__` instead of `el.constructor`

That will still not work unfortunately, for the same reason ☹️ In the content script you dont have access to the pages' JS, so `__proto__` will just return 'Element', instead of the actual WC class

**Comment by jorgecasar** [comment]
Ah, I only checked that `el.constructor.__proto__ !== HTMLUnknownElement` returns `true`.

I guess it also doesn't work with the false positives. Can you provide an example of page with false positives?

**Comment by thepassle** [comment]
>  Can you provide an example of page with false positives?

www.google.com 

You can confirm a false positive by checking `customElements.get('g-button');`, it should return undefined if its a false positive.

**Comment by LarsDenBakker** [comment]
@thepassle can it be because you're running the content script too early? Before the JS is loaded.

Some other options are to check if the element has one of the lifecycle callsbacks (`connectedCallback`, `disconnectedCallback` etc.) or if `.constructor.observedAttributes` exists.

**Comment by thepassle** [comment]
> @thepassle can it be because you're running the content script too early? Before the JS is loaded.

Thats what I thought initially, but that doesnt seem to be the case, and confirmed that its not the case with the changes here https://github.com/open-wc/locator/pull/20

Checking for callbacks/observedAttrs returns undefined

**Comment by LarsDenBakker** [comment]
did you try with a timeout? I'm not sure if the load event is late enough. Does that wait for module scripts? And wait for the code to excute?

**Comment by jorgecasar** [comment]
> [www.google.com](http://www.google.com)
> 
> You can confirm a false positive by checking `customElements.get('g-button');`, it should return undefined if its a false positive.

I don't have any `g-button` on google.com 🤦 

> did you try with a timeout? I'm not sure if the load event is late enough. Does that wait for module scripts? And wait for the code to excute?

Yes with a 1 second after window.load, and doesn't work.

```window.addEventListener('load', () => setTimeout(initSearch, 1000));```

I've checked:

```
const isAttr = el.getAttribute('is');
console.log(el.constructor.__proto__ !== HTMLUnknownElement); //true
console.log(el.constructor.__proto__ !== HTMLElement); //true
console.log(el.localName.includes('-')); //true
console.log(isAttr && isAttr.includes('-')); // null
```

We can create a demo page with all use cases and test them. I can think of:

1. Custom Element
2. Customized built-in elements
3. Regular HTMLElement
4. Unregistered element

**Comment by thepassle** [comment]
```js
    console.log(el.constructor !== HTMLUnknownElement) // true
    console.log(el.constructor !== HTMLElement) // false
    console.log(el.constructor) // HTMLElement
    console.log(el.constructor.__proto__) // Element
    console.log(el.constructor.observedAttributes) // undefined
    console.log(el.connectedCallback) // undefined
    console.log(el.localName.includes('-')) // true
```

**Comment by LarsDenBakker** [comment]
:( 

What happens when you do `document.createElement<tagname>)` ? And if you re-append that to the page? (probably a bad idea)

**Comment by jorgecasar** [comment]
Same:
```
const elem = document.createElement(el.localName);
console..log(elem.constructor); // ƒ HTMLElement()
```

**Comment by thepassle** [comment]
It looks like you should be able to access the window with a workaround...

`content_script.js`:
```js
const s = document.createElement('script');
s.innerHTML = `
setTimeout(() => {
  const allCustomElements = [/* get CEs */];

  document.dispatchEvent(new CustomEvent('foo', {
      detail: allCustomElements 
  }));
}, 1000);
`;
document.head.append(s);

document.addEventListener('foo', function({detail}) {
    console.log('all custom elements:', detail);
});
```

^ Untested, i'll try it out now

**Comment by thepassle** [comment]
Hmm you still cant send any globals back to the content script, but you can check if `customElements.get('foo-bar') !== undefined` and send back a boolean, so we'd have to move the logic of querySelecting and checking if an element is a custom element to the ACTUAL page rather than a content script (which runs besides) a page

**Comment by thepassle** [comment]
Alright, looks like I have a working implementation, would like to do some more thorough testing though. @jorgecasar would you be willing to install the extension locally and do some testing as well?

**Pull request merged**